### PR TITLE
Nice messages for https errors

### PR DESCRIPTION
**Pull request by fdbeirao**
This is an initial proposal for the strategy on how to display nice error messages for HTTPS related errors

This is related to #5487

### CODE REVIEW

**Comment by fdbeirao** [comment]
This PR brings the following messages:

Note: all files are identified as `UNKNOWN.roc`, I believe this is still a work in progress.

Note: the demo URL is hard-coded as `https://hard-coded-url-for-demo` but the errors below are independent from this URL

---

## Problem::UnsupportedEncoding("application/json")

```
── UNSUPPORTED ENCODING ────────────────────────────────────────── UNKNOWN.roc ─

I was trying to download this URL:

    https://hard-coded-url-for-demo

But the server replied with a content encoding that I do not
understand (application/json).

The supported content encodings are br, gzip and deflate

Tip: Perhaps you can check if the URL is correctly formed, or if the
server is correctly configured.
```

image

---

## Problem::MultipleEncodings("application/json, application/xml")

```
── MULTIPLE ENCODINGS ──────────────────────────────────────────── UNKNOWN.roc ─

I was trying to download this URL:

    https://hard-coded-url-for-demo

But the server replied with multiple content encodings:
application/json, application/xml.

The supported content encodings are br, gzip and deflate. However, the
server reply can only contain one.

Tip: Perhaps you can check if the URL is correctly formed, or if the
server is correctly configured.
```

image

---

## Problem::InvalidContentHash { expected: "tExpectedmxmHwHih9kHWQ7fsXtJr7W7h3425-eZFk", actual: "tE4actualmxmHwHih9kHWQ7fsXtJr7W7h3425-eZFk" }

```
── INVALID CONTENT HASH ────────────────────────────────────────── UNKNOWN.roc ─

I was able to download this URL:

    https://hard-coded-url-for-demo

I use a mechanism to detect if the file might have been tampered with.
This could happen if the server or domain have been compromised.

This is the content signature I was expecting:

    tExpectedmxmHwHih9kHWQ7fsXtJr7W7h3425-eZFk

However, this is the content signature I obtained:

    tE4actualmxmHwHih9kHWQ7fsXtJr7W7h3425-eZFk

To keep you secure, I will not execute this untrusted code.

Tip: Check if the URL is correctly formed and if this is the server
you are expecting to connect to.
```

image

---

## Problem::IoErr(io::Error::new(io::ErrorKind::Other, "oh no, an error!"))

```
── IO ERROR ────────────────────────────────────────────────────── UNKNOWN.roc ─

I was trying to download this URL:

    https://hard-coded-url-for-demo

But I encountered an IO (input/output) error:

    oh no, an error!
```

image

---

## Problem::FsExtraErr(...)

I couldn't easily instantiate an `fs_extra Error`. The report output is the same as Problem::IoErr

## Problem::HttpErr(...)

⚠ It seems like the reqwest errors are Opaque and cannot be instantiated by the user.

---

## Problem::InvalidUrl(UrlProblem::InvalidExtensionSuffix(".zip"))

```
── INVALID EXTENSION SUFFIX ────────────────────────────────────── UNKNOWN.roc ─

I was trying to download this URL:

    https://hard-coded-url-for-demo.com

However, this file's extension (.zip) is not a supported extension.

The supported extensions are .tar, .tar.gz and .tar.br

Tip: Check that you have the correct URL for this package/platform.
```

image

---

## Problem::InvalidUrl(UrlProblem::MissingTarExt)

```
── INVALID EXTENSION ───────────────────────────────────────────── UNKNOWN.roc ─

I was trying to download this URL:

    https://hard-coded-url-for-demo.com

However, this file's extension is not .tar.

The supported extensions are .tar, .tar.gz and .tar.br

Tip: Check that you have the correct URL for this package/platform.
```

image

---

## Problem::InvalidUrl(UrlProblem::InvalidFragment("filename.exe")),

```
── INVALID FRAGMENT ────────────────────────────────────────────── UNKNOWN.roc ─

I was trying to download this URL:

    https://hard-coded-url-for-demo.com

However, this URL's fragment (the part after #) is not valid. When
present, the fragment must point to an existing .roc file inside the
package. Also, the filename can't be empty so a fragment of #.roc
would also not be valid. This is the problematic fragment I
encountered: 

    filename.exe

Tip: Check that the fragment points to an existing .roc file inside
the package. You can download this package and inspect it locally.
```

image

---

## Problem::InvalidUrl(UrlProblem::MissingHash)

```
── MISSING PACKAGE HASH ────────────────────────────────────────── UNKNOWN.roc ─

I was trying to download this URL:

    https://hard-coded-url-for-demo.com

I use a mechanism to detect if the file might have been tampered with.
This could happen if the server or domain have been compromised.

The way this works is that the name of the file is the cryptographic
hash of the contents of the file itself. If someone would tamper with
the file, I could notify and protect you. However, I could not find
the expected hash on the URL above, so I cannot apply this
tamper-check.

Tip: Check that you have the correct URL for this package/platform.
Here is an example of how such a hash looks like:
tE4xS_zLdmmxmHwHih9kHWQ7fsXtJr7W7h3425-eZFk
```

image

---

## Problem::InvalidUrl(UrlProblem::MissingHttps)

```
── HTTPS MANDATORY ─────────────────────────────────────────────── UNKNOWN.roc ─

I was trying to download this URL:

    https://hard-coded-url-for-demo.com

For your security, I will only attempt to download files from servers
which use the https protocol.

Tip: Check that you have the correct URL for this package/platform.
```

image

---

## Problem::InvalidUrl(UrlProblem::MisleadingCharacter)

```
── MISLEADING CHARACTERS DETECTED ──────────────────────────────── UNKNOWN.roc ─

I was trying to download this URL:

    https://hard-coded-url-for-demo.com

I have found one or more potentially misleading characters in this
URL. Misleading characters are characters that look like others but
aren't the same. The following characters are classified as
misleading: @, ⁄ (unicode 2044), ∕ (unicode 2215), ／ (unicode
FF0F) and ⧸ (unicode 29F8). 

If you have a use-case for any of these characters we would like to
hear about it. Reach out on
https://github.com/roc-lang/roc/issues/5487

Tip: Check that you have the correct URL for this package/platform.
```

image

---

## Problem::DownloadTooBig(123_456_789)

```
── FILE TOO LARGE ──────────────────────────────────────────────── UNKNOWN.roc ─

I was trying to download this URL:

    https://hard-coded-url-for-demo.com

But the server stated this file is 123456789 bytes in size. This is
larger that the maximum size I can handle (around 32 GB).

Tip: Check that you have the correct URL for this package/platform. If
you do, you should contact the package/platform's author and notify
them about this issue.
```

image

---

While doing some final tests for this PR I notice what when I run my `cargo run dev ../hacking/5487/main.roc` the code seems to have a slightly different execution path from what I was expecting, so I'll investigate that further to plug these nicer error messages also there. 😅

image

**Comment by fdbeirao** [comment]
Managed to get the error messages also when downloading the packages, but the reflow logic doesn't seem to be collaborating. I'm struggling a bit with reflowing strings, due to my unfamiliarity with `String` versus `&str` in rust 😰

**Comment by Anton-4** [comment]
This is looking good @fdbeirao, feel free to ask for help on zulip for specific issues with reflowing or `String` vs `&str`.

File: crates/reporting/src/report.rs [hunk 1]
```diff
@@ -1075,6 +1078,468 @@ where
...
+                alloc
+                    .string((&url).to_string())
+                    .annotate(Annotation::Url)
+                    .indent(4),
+                alloc.concat([
+                    alloc.reflow(r"However, this URL's fragment (the part after #) "),
+                    alloc.reflow(r"is not valid. When present, the fragment must point to "),
+                    alloc.reflow(r"an existing "),
+                    alloc.keyword(r".roc"),
+                    alloc.reflow(r" file inside the package. Also, the filename can't be empty "),
```
Comment (rtfeldman) [inline crates/reporting/src/report.rs:1381]:
```suggestion
                    alloc.reflow(r" file inside the package. Also, the filename can't be empty, "),
```

**Review by rtfeldman — COMMENTED** [review]
These look awesome!!! 😍 

I added some suggested wording tweaks, but overall this is fantastic - thank you so much @fdbeirao!

**Review by rtfeldman — COMMENTED** [review]
These look awesome!!! 😍 

I added some suggested wording tweaks, but overall this is fantastic - thank you so much @fdbeirao!

**Comment by fdbeirao** [comment]
I couldn't get long error messages to reflow. I think that this PR brings value on its own and that we can tackle proper reflowing of longer messages separately. So, if approved, let's merge this one 👍

**Review by ayazhafiz — APPROVED** [review]
We should add some tests, but that can be done in the future. Thanks!

**Pull request merged**